Implement busio.I2C

dhalbert opened the pull request:
* Added `asf4_conf/samd*/hpl_sercom_config.h`
* Adjusted clocks in `peripheral_clk_config.h.`
* Put some frozen libs back in CPX for testing.
* Implement common-hal I2C.
* Re-enable resetting SERCOMs in `port_init()`, so SERCOM's get reset on soft reboot.
* Add `samd*_peripherals.h` in parallel with `samd*_pins.h` for common
  functions and data.
* Store SERCOM index in pins table for convenience.
* Canonicalize some `#include` guard names in various .h files.

tannewt commented on ports/atmel-samd/supervisor/port.c:
```diff
@@ -167,21 +167,24 @@ safe_mode_t port_init(void) {
+#endif
+        sercom_instances[i]->SPI.CTRLA.bit.SWRST = 1;
+        sercom_instances[i]->I2CM.CTRLA.bit.ENABLE = 0;
+        sercom_instances[i]->I2CM.CTRLA.bit.SWRST = 1;
```
I don't think you need these three things. SPI.CTRLA.bit.SWRST is the same memory location as I2CM.CTRLA.bit.SWRST because SPI and I2CM are members of a union. Instead, I think you had "in use" errors because the reset takes time and the M4 is faster than the M0. You can wait on the state of SYNCBUSY.bit.SWRST. It'll be 0 when the reset is finished.

dhalbert replied:
Union thing understood. I didn't read the register defns carefully enough.
The weird thing about the ENABLE and SWRST is that I got "pin in use" errors after a soft reboot when lines 183 and 184 were reversed. These errors were many seconds after the reboot: I was testing in the REPL. I'll test this again.

dhalbert replied:
I thought maybe SWRST was ENABLE-protected. It says it isn't.

tannewt commented on ports/atmel-samd/common-hal/busio/I2C.c:
```diff
@@ -71,28 +70,36 @@ void common_hal_busio_i2c_construct(busio_i2c_obj_t *self,
+    // Set up I2C clocks on sercom.
+    sercom_clock_init(sercom, sercom_index);
+
+    if (i2c_m_sync_init(&self->i2c_desc, sercom) != ERR_NONE) {
```
The reliance of these new `_init` methods on arrays of `_configuration` structs is really annoying. Is the compiler smart enough to inline the values from the arrays or does it always take code space? I'd be tempted to init them ourselves to avoid the configuration structs and the need for the CONF values at all.

dhalbert replied:
`sercom_clock_init` is new by me (it's in `samd*1_periphals.c`). I do need the clock lookup table (maybe not the SLOW), because the numbers are random.

I think we could get rid of the `_i2cms` configuration table in `hp_sercom.c`, but it means changing the semantics of the library somewhat. I was trying to make minimal code changes to ASF4 aside from fixing bugs, so that we could merge more easily later.

The compiler can't inline the values because they're looked up by an index computed at runtime.

dhalbert commented on ports/atmel-samd/common-hal/busio/I2C.c:
```diff
@@ -103,37 +110,40 @@ void common_hal_busio_i2c_deinit(busio_i2c_obj_t *self) {
 
 void common_hal_busio_i2c_configure(busio_i2c_obj_t *self,
         uint32_t baudrate, uint8_t polarity, uint8_t phase, uint8_t bits) {
+    // Not implemented (?).
```
I'll delete.

dhalbert commented:
I fixed the unnecessary SERCOM reset and removed the unused routine. I have some ideas for getting rid of the configuration tables. But are you willing to defer that for a little while and take the current implementation so people can start testing with I2C devices? I'll come back to the configuration stuff when I start implementing the next part of busio.
